cupy/fft/_cache.pyx

```diff
@@ -0,0 +1,612 @@
+# distutils: language = c++
+
+from cupy_backends.cuda.api cimport runtime
+from cupy.cuda cimport device
+from cupy.cuda cimport memory
+
+import threading
+
+from cupy import _util
+from cupy.cuda import cufft
+
+
+#####################################################################
+# Internal implementation                                           #
+#####################################################################
+
+cdef object _thread_local = threading.local()
+
+
+cdef class _ThreadLocal:
+
+    cdef list per_device_cufft_cache
+
+    def __init__(self):
+        cdef int i
+        self.per_device_cufft_cache = [
+            None for i in range(runtime.getDeviceCount())]
+
+    @staticmethod
+    cdef _ThreadLocal get():
+        cdef _ThreadLocal tls
+        tls = getattr(_thread_local, 'tls', None)
+        if tls is None:
+            tls = _ThreadLocal()
+            setattr(_thread_local, 'tls', tls)
+        return tls
+
+
+cdef inline Py_ssize_t _get_plan_memsize(plan, int curr_dev=-1) except -1:
+    cdef Py_ssize_t memsize = 0
+    cdef memory.MemoryPointer ptr
+    cdef int dev
+
+    # work_area could be None for "empty" plans...
+    if plan is not None and plan.work_area is not None:
+        if plan.gpus is not None:
+            # multi-GPU plan
+            if curr_dev == -1:
+                curr_dev = runtime.getDevice()
+            for dev, ptr in zip(plan.gpus, plan.work_area):
+                if dev == curr_dev:
+                    memsize = <Py_ssize_t>(ptr.mem.size)
+                    break
+            else:
+                raise RuntimeError('invalid multi-GPU plan')
+        else:
+            # single-GPU plan
+            ptr = plan.work_area
+            memsize = <Py_ssize_t>(ptr.mem.size)
+
+    return memsize
+
+
+cdef class _Node:
+    # Unfortunately cython cdef class cannot be nested, so the node class
+    # has to live outside of the linked list...
+
+    # data
+    cdef readonly tuple key
+    cdef readonly object plan
+    cdef readonly Py_ssize_t memsize
+    cdef readonly list gpus
+
+    # link
+    cdef _Node prev
+    cdef _Node next
```

Comment:
It causes a circular reference, so some memory spaces may not be free unexpectedly.
How about using a singly-linked list instead of the current double-linked list? It may cost more time to eject a node from the linked list, but it seems to be small enough compared to the time to make a new plan.

Reply:
Perhaps using Python's primitive list may be simpler than a singly-linked list.

Reply:
Hi @asi1024, could you help me understand better about the circular reference issue? Is it because when ejecting a node, I didn't clear out the node's `prev` and `next`?

Reply:
If `_LinkedList` is deleted for some reason (not when ejecting a node), its nodes will not be deleted immediately because they have circular references.
One solution for this issue is changing the data structure to a singly-linked list by deleting a pointer (maybe `next`?). It costs linear time to eject an element because we should trace the linked list from the other side, but I think it is ignorable compared to the cpu-time to make a new plan.

Reply:
It sounds like we could also just add a destructor for `_LinkedList`, walk through the nodes, and delete them in order?

Reply:
I think both of them work correctly. I prefer singly-linked list or Python's primitive list for maintainability, but adding a destructor is also OK if the difference from the current implementation is small.

Reply:
Sorry, now I notice that I was mentioned at issue #3999.

Reply:
Yeah, looks like we can't do `__del__` here. I added a `clear()` routine and invoked it manually when needed, see 8d0705c.

Reply:
@asi1024 Give me some time to improve the `clear()` routine. A colleague of mine reminded me that `weakref` is exactly for such purposes: https://docs.python.org/3/library/weakref.html#comparing-finalizers-with-del-methods